### PR TITLE
LLM tech report performance analysis

### PR DESCRIPTION
**Pull request by yieldthought**

### CODE REVIEW

File: tech_reports/LLMs/llms.md [hunk 1]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+2. **Host API** - most of your ttnn calls will be immediately dispatched onto multiple C++ threads for further processing before anything is done to the hardware. You are generally not in control of any overheads in this part of the stack.
+3. **Host-device communications** - data is heavy and moving it is to be avoided. PCIe bandwidth and latency is not negligible at the kinds of speeds we want to run models at. In addition, Tenstorrent converts most data into tiles of 32x32 elements for faster processing. Tilizing and untilizing data takes time and should be performed on-device wherever possible, but you will have to specify that as we’ll see in this section.
+4. **Device dispatch** - we can measure the gap between one op finishing and the next starting. At time of writing the lower limit of this is single-digit microseconds and there is work underway to reduce it to zero. However, for various reasons you might see much higher dispatch times, most notably if there are a lot of runtime arguments to a function or if something else is happening in between calls.
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:98]: I think this work has been partially pushed to main already. So, maybe mention that dispatch will be overlapped with previous op execution whenever possible.

File: tech_reports/LLMs/llms.md [hunk 2]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+
+We will dive into all of these in detail - how we like to measure each section and tips and tricks to optimize it.
+However, the high-order bit we should consider first is this: should you use tracing?
+
+#### What is tracing and when (not) to use it
+
+Check out the [Metal Trace guide](https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/AdvancedPerformanceOptimizationsForModels/AdvancedPerformanceOptimizationsForModels.md) for background on this. Essentially it lets you record a single pass of your model and stores the list of commands and buffers used on-device. You can then execute that trace in a single command with no additional work performed on the host. This eliminates all the overhead in stages 1-4 (you are still responsible for transferring any data needed to and from device, but host-device transfer of commands is eliminated).
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:106]: I think tracing does not eliminate device dispatch. I am not 1-4 also includes device dispatch.

File: tech_reports/LLMs/llms.md [hunk 3]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+
+We typically use tracing for the decode pass of LLMs but not the prefill pass. The main reasons for this are linked to tracing’s key limitation:
+
+* You cannot allocate or deallocate tensors during a trace. When executing a trace every buffer will be the same size every time.
+
+This doesn’t fit well with prefill, in which the sequence length and so matmul row counts will likely change each time. For decode it’s no problem (see the sections on kv-cache and paging for how we handle those with tracing). Conveniently, in prefill we have big operations in the millisecond plus range which the host can usually dispatch fast enough that it doesn’t matter. Decode, with a comparatively small batch size, is another story. There we iterate through the entire model in like 10ms with microsecond-length op times where we cannot afford to wait for a CPU, the whims of linux process scheduling or anything else but the speed at which electrons coruscate from DRAM and the NoC through our cores and out again.
+**TL;DR: for decode mode you won’t have to worry about 1-3 but for prefill mode you will.** We will cover everything anyway.
+
+#### What is async mode and when (not) to use it
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:115]: I added a separate async mode section (4.2) before Performance Analysis.

File: tech_reports/LLMs/llms.md [hunk 4]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+4. **Device dispatch** - we can measure the gap between one op finishing and the next starting. At time of writing the lower limit of this is single-digit microseconds and there is work underway to reduce it to zero. However, for various reasons you might see much higher dispatch times, most notably if there are a lot of runtime arguments to a function or if something else is happening in between calls.
+5. **Device op performance** - how long it takes the hardware to run a given operation. Ideally we want this to be limited either by DRAM bandwidth or math throughput and for larger ops both of these are generally achievable. Doing so is mostly about how the data is placed (DRAM vs L1, sharded vs interleaved) and how the compute kernels are configured (process more than one tile at once and use smaller data formats).
+
+We will dive into all of these in detail - how we like to measure each section and tips and tricks to optimize it.
+However, the high-order bit we should consider first is this: should you use tracing?
+
+#### What is tracing and when (not) to use it
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:104]: I added a separate Tracing section (4.1) before Performance Analysis.

File: tech_reports/LLMs/llms.md [hunk 5]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+There are two main contributors to op-to-op gap: **host time** and **dispatch time**.
+
+* **Host time** is optimized in steps 1-3. If you are already using tracing or are using async mode and have ensured that your python thread is dispatching faster than the device is generating outputs then this has already been minimized.
+* **Dispatch time** is mostly out of your hands, but as an example it is influenced by the number of runtime args a kernel uses.
+    * You can examine the source code for any kernel with high op-to-op latency and see if you can convert some of the runtime args into compile-time args for your use case.
+    * You can fuse multiple ops into a single kernel. Examples where this was well worthwhile in the past include `LayerNorm` and `ScaledDotProductAttentionDecode`.
+
+Typically tracing reduces the op-to-op gap below 6us and as of November 2024 there are roadmap plans to reduce this to zero, so as long as your ops are below this level your opportunities for optimization here are limited.
+
+##### Overall op performance advice
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:242]: I added a separate Misc. Performance Optimizations section (4.7) after Performance Analysis. This text is perfect for that.

File: tech_reports/LLMs/llms.md [hunk 6]
```diff
@@ -86,7 +87,231 @@ Authors:
...
+
+> **Note:** If you see errors whilst running tracy, try this device-only profiling process instead: run with `TT_METAL_DEVICE_PROFILER=1 pytest path/to/test.py` and after the run completes run `tt_metal/tools/profiler/process_ops_logs.py --date` to generate the CSV file.
+
+This CSV file contains a wealth of information recorded from all the device during program execution. To summarize it into a more human-readable format we run the `perf_report.py` tool:
+
+```bash
+python models/perf/perf_report.py OPS_CSV_FILE
+```
+
+The [documentation for this tool](https://github.com/tenstorrent/tt-metal/tree/main/models/perf) describes how to use it to select specific ranges of operations. For device performance we strongly recommend looking at a single layer. You can do this by using `--id-range` or by changing your test to run only a single layer of the model (recommended).
```
Comment (uaydonat) [inline tech_reports/LLMs/llms.md:214]: Where is the documentation for perf_report.py. The link only contains the source code for the scripts.

File: tech_reports/LLMs/llms.md (hunk 1 above)
Comment (yieldthought) [reply tech_reports/LLMs/llms.md:98]: I don't see any zero dispatch times, they're all 4us with tracing enabled on main atm.

File: tech_reports/LLMs/llms.md (hunk 4 above)
Comment (yieldthought) [reply tech_reports/LLMs/llms.md:104]: Moved to there.

File: tech_reports/LLMs/llms.md (hunk 5 above)
Comment (yieldthought) [reply tech_reports/LLMs/llms.md:242]: Moved to that section

**Comment by yieldthought** [comment]
@mtairum Review comments all addressed, this is ready for merge into your master branch now.

**Review by davorchap — APPROVED** [review]

**Comment by davorchap** [comment]
Please merge to main, a customer would like to use it

**Review by tt-rkim — APPROVED** [review]

**Pull request merged**